### PR TITLE
Fix multi-file download 

### PR DESCRIPTION
**Pull request by sfc-gh-ext-simba-ll**
issue [175](https://github.com/snowflakedb/snowflake-sdks-drivers-issues-teamwork/issues/175)
The cause of the issue: uploads uses the metadata from the first file to encrypt, and it also reads from the first file to decrypt for downloads. This method works when uploading files all together and then download them all together. But when uploading files one by one, it will use different encryption material, then it cannot download all the files together. That's why only the first file succedded to download. The fix is to set the download to read the encryption material from each file and make sure all the files has its own encryption material when uploading it.

### CODE REVIEW

**Review by sfc-gh-igarish — APPROVED** [review]
LGTM

File: test/integration/testPutGet.js [hunk 1]
```diff
@@ -657,3 +658,329 @@ describe('PUT GET test with GCS_USE_DOWNSCOPED_CREDENTIAL', function ()
...
+
+    var putQuery1 = `PUT file://${tmpFile1.name} @${DATABASE_NAME}.${SCHEMA_NAME}.%${TEMP_TABLE_NAME}`;
+    // Windows user contains a '~' in the path which causes an error
+    if (process.platform == "win32")
+    {
+      var fileName = tmpFile1.name.substring(tmpFile1.name.lastIndexOf('\\'));
+      putQuery1 = `PUT file://${process.env.USERPROFILE}\\AppData\\Local\\Temp\\${fileName} @${DATABASE_NAME}.${SCHEMA_NAME}.%${TEMP_TABLE_NAME}`;
+    }
+
+    var putQuery2 = `PUT file://${tmpFile2.name} @${DATABASE_NAME}.${SCHEMA_NAME}.%${TEMP_TABLE_NAME}`;
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testPutGet.js:702]: can we create local putFileQuery method to generate proper queries with win32 specific behaviour written and commented in one place?

File: test/integration/testPutGet.js [hunk 2]
```diff
@@ -657,3 +658,329 @@ describe('PUT GET test with GCS_USE_DOWNSCOPED_CREDENTIAL', function ()
...
+      var dirName = tmpDir.substring(tmpDir.lastIndexOf('\\') + 1);
+      getQuery = `GET @${DATABASE_NAME}.${SCHEMA_NAME}.%${TEMP_TABLE_NAME} file://${process.env.USERPROFILE}\\AppData\\Local\\Temp\\${dirName}`;
+    }
+
+    async.series(
+      [
+        function (callback)
+        {
+          // Create temp table
+          testUtil.executeCmd(connection, createTable, callback);
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testPutGet.js:727]: createTable is used only here - can we have this variable local in this lambda?

File: test/integration/testPutGet.js [hunk 3]
```diff
@@ -657,3 +658,329 @@ describe('PUT GET test with GCS_USE_DOWNSCOPED_CREDENTIAL', function ()
...
+        {
+          // Upload file
+          var statement = connection.execute({
+            sqlText: putQuery1,
+            complete: function (err, stmt, rows)
+            {
+              var stream = statement.streamRows();
+              stream.on('error', function (err)
+              {
+                done(err);
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testPutGet.js:739]: shouln't we call `callback(err)`?

File: test/integration/testPutGet.js [hunk 4]
```diff
@@ -657,3 +658,329 @@ describe('PUT GET test with GCS_USE_DOWNSCOPED_CREDENTIAL', function ()
...
+              {
+                callback();
+              });
+            }
+          });
+        },
+        function (callback)
+        {
+          // Upload file
+          var statement = connection.execute({
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testPutGet.js:759]: this execution is the same as the one above but put's different file? can we create one method for put?

File: test/integration/testPutGet.js [hunk 5]
```diff
@@ -657,3 +658,329 @@ describe('PUT GET test with GCS_USE_DOWNSCOPED_CREDENTIAL', function ()
...
+        {
+          // Run GET command
+          var statement = connection.execute({
+            sqlText: getQuery,
+            complete: function (err, stmt, rows)
+            {
+              var stream = statement.streamRows();
+              stream.on('error', function (err)
+              {
+                done(err);
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testPutGet.js:793]: shouldn't it be `callback(err)`?

File: test/integration/testPutGet.js [hunk 6]
```diff
@@ -657,3 +658,329 @@ describe('PUT GET test with GCS_USE_DOWNSCOPED_CREDENTIAL', function ()
...
+              {
+                callback();
+              });
+            }
+          });
+        },
+        function (callback)
+        {
+          // Remove files from staging
+          testUtil.executeCmd(connection, removeFile, callback);
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testPutGet.js:825]: could `removeFile` be local variable?

File: test/integration/testPutGet.js [hunk 7]
```diff
@@ -657,3 +658,329 @@ describe('PUT GET test with GCS_USE_DOWNSCOPED_CREDENTIAL', function ()
...
+        },
+        function (callback)
+        {
+          // Drop temp table
+          testUtil.executeCmd(connection, dropTable, callback);
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testPutGet.js:830]: could `dropTable` be local variable?

File: test/integration/testPutGet.js [hunk 8]
```diff
@@ -657,3 +658,329 @@ describe('PUT GET test with GCS_USE_DOWNSCOPED_CREDENTIAL', function ()
...
+      ],
+      done
+    );
+  });
+
+  it('testUploadMultifiles', function (done)
+  {
+    const count = 5;
+    var tmpFiles = [];
+    var createTable = `create or replace table ${TEMP_TABLE_NAME} (${COL1} STRING, ${COL2} STRING, ${COL3} STRING)`;
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testPutGet.js:855]: could we create the table once in before method for both tests?

File: test/integration/testPutGet.js [hunk 9]
```diff
@@ -657,3 +658,329 @@ describe('PUT GET test with GCS_USE_DOWNSCOPED_CREDENTIAL', function ()
...
+    var removeFile = `REMOVE @${DATABASE_NAME}.${SCHEMA_NAME}.%${TEMP_TABLE_NAME}`;
+    var dropTable = `DROP TABLE IF EXISTS ${TEMP_TABLE_NAME}`;
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testPutGet.js:857]: could we cleanup table in after method for both tests?

File: test/integration/testPutGet.js [hunk 10]
```diff
@@ -657,3 +658,329 @@ describe('PUT GET test with GCS_USE_DOWNSCOPED_CREDENTIAL', function ()
...
+
+    // Create a tmp folder for downloaded files
+    var tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'get'));
+
+    var fileSize;
+
+    var getQuery = `GET @${DATABASE_NAME}.${SCHEMA_NAME}.%${TEMP_TABLE_NAME} file://${tmpDir}`;
+    // Windows user contains a '~' in the path which causes an error
+    if (process.platform == "win32")
```
Comment (sfc-gh-dprzybysz) [inline test/integration/testPutGet.js:866]: maybe let's create one method to generate proper file path depending on the platform?

**Comment by sfc-gh-dprzybysz** [comment]
Let's remember to add jira issue to commit message on squash

**Pull request merged**